src/timemachine.c

```diff
@@ -0,0 +1,613 @@
+ * \retval TM_ECODE_FAILED on serious error
+ */
+static TmEcode TimeMachineProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
+                 PacketQueue *postpq)
+{    
+    TimeMachineFlow *flow;
+    TimeMachineHeap *heap;
+    TimeMachineMemPool *mem_pool;
+    TimeMachinePacket *packet;
+    struct pcap_pkthdr pkthdr;
+    struct timeval current_time;
+    
+    TimeGet(&current_time);
+    
+    TimeMachineThreadData *td = (TimeMachineThreadData *)thread_data;
+    TimeMachineData *tm = td->tm_data;
+    
+    /* make sure packet has real data */
+    if (GET_PKT_PAYLOAD_LEN(p) == 0) {
+        return TM_ECODE_OK;
+    }
+    
+    /* check whether a corresponding flow already exists */
+    if (!(flow = TimeMachineFlowLookup(td->flows, p))) {
+        flow = TimeMachineFlowNew(td->flows, p);       
+        td->flow_count++;
+        flow->td = td;
+    }
+       
+    /* if this packet has an alert but flow doesn't have a output, create one */
+    if (p->alerts.cnt > 0 && flow->output == NULL) {
+        flow->output = TimeMachineOutputNew(flow);
+        TAILQ_INSERT_TAIL(&td->outputs, flow->output, next);
+        td->output_count += 1;
+    }
+     
+     /* create the packet header info */
+    pkthdr.ts.tv_sec = p->ts.tv_sec;
+    pkthdr.ts.tv_usec = p->ts.tv_usec;
+    pkthdr.caplen = GET_PKT_LEN(p);
+    pkthdr.len = GET_PKT_LEN(p);
+                                         
+    /* dump the traffic if this has already been marked as a output */
+    if (flow->output != NULL) {
+        pcap_dump((u_char*)flow->output->output_file, &pkthdr, GET_PKT_DATA(p));
+        fflush(flow->output->output_file);
+        memcpy(&flow->output->updated, &current_time, sizeof(struct timeval));
+        
+        /* make this the most recently accessed output */
+        TAILQ_REMOVE(&td->outputs, flow->output, next);
+        TAILQ_INSERT_TAIL(&td->outputs,flow->output, next);
+        
+        return TM_ECODE_OK;
+    } 
+    
+    /* find the queue the packet should fall in */
+    TAILQ_FOREACH(heap, &td->heaps, next) {
+        if (pkthdr.caplen <= heap->conf->max_packet_size) {
+            break;
+        }
+    }
+   
+    /* see if something is already available in the unusued mem pool */
+    if (TAILQ_EMPTY(&heap->unused_mem_pools)) {
+
+        /* first check if the heap can expand, if so then expand it */
+        SCMutexLock(&tm->tm_lock);   
+        if (TimeMachineHeapCanExpand(tm)) {
+            TimeMachineHeapExpand(tm, heap);
+            SCMutexUnlock(&tm->tm_lock);
+        }
+        /* couldn't expand, need to remove the first entry */
+        else {
+            SCMutexUnlock(&tm->tm_lock);
+           
+            TimeMachineFlow* rem_flow;                                                                                                                                                                                                                                        
```

Comment:
Not shown in the diff on github, but there are a few lines of extra spaces here.